doc/tutorial.md

```diff
@@ -0,0 +1,594 @@
+```
+
+Then, the `zebra/ospfd` part. Firstly, to run the OSPF protocol between `DPVS` server and wan-side L3-switch, with the "inter-connection network" (here is `172.10.0.2/30`). For `DPVS`, we set the inter-connection IP on `dpdk1.kni`.
+
+> Assuming `quagga` package is installed, if not, pls use 'yum' (CentOS) or 'apt-get' (Ubuntu) to install it. After installed, you should have `zebra` and `ospfd`, as well as their config files.
+
+```bash
+$ ip link set dpdk1.kni up
+$ ip addr add 172.10.0.2/30 dev dpdk1.kni
+$ ip addr add 123.1.2.3/32 dev dpdk1.kni # add VIP to kni for ospfd
+$ ip route add default via 172.10.0.1 dev dpdk1.kni
+```
+
+> VIP should be add to kni device, to let ospfd to publish it.
+
+Check if inter-connection works by `ping` switch.
+
+```bash
+$ ping 172.10.0.1
+PING 172.10.0.1 (172.10.0.1) 56(84) bytes of data.
+64 bytes from 172.10.0.1: icmp_seq=1 ttl=255 time=2.19 ms
+```
+
+Now let's config `zebra` and `ospfd`. Nothing special for `zebra`, just use it with the default configuration.
+
+```bash
+$ cat /etc/quagga/zebra.conf  # may installed to other path
+! -*- zebra -*-
+!
+! zebra sample configuration file
+!
+! Id: zebra.conf.sample,v 1.1 2002/12/13 20:15:30 paul Exp $
+!
+hostname localhost.localdomain
+password ****
+enable password ****
+
+log file /var/log/quagga/zebra.log
+service password-encryption
+```
+
+For `ospfd`, these parameters need be set:
+
+* interface: it's WAN interface `dpdk1.kni`
+* route-id: not that significant, just use the LAN IP.
+* network: which network to broadcasting
+  - the inter-connection network `172.10.0.0/30`, and
+  - the VIP `123.1.2.3/32`.
+* area-ID: should be same with switch, here is `0.0.0.0` for example.
+* Other parameters, like "p2p", "authentication", ... they must be consistent with Switch.
+
+```bash
+$ cat /etc/quagga/ospfd.conf   # may installed to other path
+log file /var/log/quagga/ospf.log
+log stdout
+log syslog
+password ****
+enable password ****
+interface dpdk1.kni      # should be wan-side kni device
+ip ospf hello-interval 10
+ip ospf dead-interval 40
+router ospf
+ospf router-id 192.168.100.200 # just use LAN IP
+log-adjacency-changes
+auto-cost reference-bandwidth 1000
+network 172.10.0.0/30 area 0.0.0.0 # announce inter-connection network
+network 123.1.2.3/32 area 0.0.0.0 # announce VIP
+```
+
+Note `OSPF` must also be configured on l3-switch. This Tutorial is not about OSPF's configuration, so no more things about switch here.
+
+Now start `zebra` and `ospfd`:
+
+```bash
+service restart zebra
+service restart ospfd
+```
+
+Hopefully (if `OSPF` works), the VIP is accessible by client:
+
+```bash
+client: curl 123.1.2.3
+```
+
+<a id='fnat-keepalive'/>
+
+## Full-NAT with Keepalived (one-arm)
+
+This is an example for FullNAT used in internal network (LAN). `Keepalived`  (*`DPVS` modified version*) is used for to make DPVS works as *Master/Backup* model.
+
+![fnat-keepalive](pics/fnat-keepalived.png)
```

Comment:
diagram: the lines better to connect dpdk0.kni and VRRP switch?

Reply:
the line is for data forwarding.